**ASSIGNMENT 4**

1. Group Name: 110
   Member Names: Axel Tang
   Member Student Numbers: 300164095
   Report Title: CSI4106 A4 - Classificiation Empirical Study | Text Classification

**Derived Datasets**

This notebook is a starting point for Assignment 4. In this assignment, you will perform a classification empirical study. This notebook will help you to create derived datasets in Section 2 of the assignment.

In [240]:
#let's start by installing spaCy
import warnings
warnings.simplefilter('ignore')
!pip install spacy



[notice] A new release of pip available: 22.2.2 -> 23.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [241]:
import spacy
import pandas as pd
import numpy as np

You have been given a list of datasets in the assignment description. Choose one of the datasets and provide the link below and read the dataset using pandas. You should provide a link to your own Github repository even if you are using a reduced version of a dataset from your TA's repository.

add description of the dataset and your justification of the choices made to obtain the derived datasets

In [242]:
#Load the dataset you chose.
# Make sure the Notebook can load your dataset, just like previous assignments.

# url = 'https://raw.githubusercontent.com/baharin/CSI4106-Assignment4-Datasets/main/reduced_file_cnnnews.csv'
# url = 'https://raw.githubusercontent.com/baharin/CSI4106-Assignment4-Datasets/main/reduced_drugsComTest_raw_fiveclasses.csv'
#provide the link to the raw version of dataset. You *need* to provide a link to *your own* github repository. DO NOT use the link that is provided as an example.
url = 'https://raw.githubusercontent.com/AxelTWC/CSI_AI-Assignment4-CSV/main/reduced_file_cnnnews.csv'




In [243]:
print(url)
data = pd.read_csv(url)

https://raw.githubusercontent.com/AxelTWC/CSI_AI-Assignment4-CSV/main/reduced_file_cnnnews.csv


In [244]:
data.head()

ID                                              Title  \
0   1  Candy factory didn't evacuate concerned worker...   
1   2  Baltimore police ask for public's help identif...   
2   3  An arrest warrant has been issued for a suspec...   
3   4  115 improperly stored human remains found in C...   
4   5  Bronx day care provider and 2 others indicted ...   

                                         Description  \
0  An eastern Pennsylvania candy factory didn’t e...   
1  Two shooters were involved in an attack at Mor...   
2  Authorities in Pennsylvania say they have issu...   
3  An investigation into more than 115 bodies fou...   
4  A Bronx day care provider, her husband and his...   

                                                Body  \
0  An eastern Pennsylvania candy factory didn’t e...   
1  Two shooters were involved in an attack at Mor...   
2  Authorities in Pennsylvania say they have issu...   
3  An investigation into more than 115 bodies fou...   
4  A Bronx day care provider, her husband and his...   

                                            Keywords Theme  \
0  accident investigations, accidents, accidents,...    us   
1  baltimore, brand safety-nsf crime, brand safet...    us   
2  arrest warrants, arrests, brand safety-nsf cri...    us   
3  brand safety-nsf death, brand safety-nsf sensi...    us   
4  brand safety-nsf crime, brand safety-nsf death...    us   

                                                Link  
0  https://edition.cnn.com/2023/10/06/us/pennsylv...  
1  https://edition.cnn.com/2023/10/06/us/morgan-s...  
2  https://edition.cnn.com/2023/10/06/us/josh-kru...  
3  https://edition.cnn.com/2023/10/06/us/colorado...  
4  https://edition.cnn.com/2023/10/05/us/bronx-da...

This is where you create the NLP pipeline. load() will download the correct model (English).

In [245]:
!python -m spacy download en_core_web_sm #<- Download spacy packages.
nlp = spacy.load("en_core_web_sm")

The system cannot find the file specified.


Applying the pipeline to every sentences creates a Document where every word is a Token object.

Doc: https://spacy.io/api/doc

Token: https://spacy.io/api/token

In [246]:
#Apply nlp pipeline to the column that has your sentences.
data['tokenized'] = data['Title'].apply(nlp)

In [247]:
data.head()

ID                                              Title  \
0   1  Candy factory didn't evacuate concerned worker...   
1   2  Baltimore police ask for public's help identif...   
2   3  An arrest warrant has been issued for a suspec...   
3   4  115 improperly stored human remains found in C...   
4   5  Bronx day care provider and 2 others indicted ...   

                                         Description  \
0  An eastern Pennsylvania candy factory didn’t e...   
1  Two shooters were involved in an attack at Mor...   
2  Authorities in Pennsylvania say they have issu...   
3  An investigation into more than 115 bodies fou...   
4  A Bronx day care provider, her husband and his...   

                                                Body  \
0  An eastern Pennsylvania candy factory didn’t e...   
1  Two shooters were involved in an attack at Mor...   
2  Authorities in Pennsylvania say they have issu...   
3  An investigation into more than 115 bodies fou...   
4  A Bronx day care provider, her husband and his...   

                                            Keywords Theme  \
0  accident investigations, accidents, accidents,...    us   
1  baltimore, brand safety-nsf crime, brand safet...    us   
2  arrest warrants, arrests, brand safety-nsf cri...    us   
3  brand safety-nsf death, brand safety-nsf sensi...    us   
4  brand safety-nsf crime, brand safety-nsf death...    us   

                                                Link  \
0  https://edition.cnn.com/2023/10/06/us/pennsylv...   
1  https://edition.cnn.com/2023/10/06/us/morgan-s...   
2  https://edition.cnn.com/2023/10/06/us/josh-kru...   
3  https://edition.cnn.com/2023/10/06/us/colorado...   
4  https://edition.cnn.com/2023/10/05/us/bronx-da...   

                                           tokenized  
0  (Candy, factory, did, n't, evacuate, concerned...  
1  (Baltimore, police, ask, for, public, 's, help...  
2  (An, arrest, warrant, has, been, issued, for, ...  
3  (115, improperly, stored, human, remains, foun...  
4  (Bronx, day, care, provider, and, 2, others, i...

A Token object has many attributes such as part-of-speech (pos_), lemma (lemma_), etc. Take a look at the documentation to see all attributes.

The following function is an example on how you can fetch a specific pos tagging from a sentence. We return the lemmatization because we only want the infinitive word.

In [248]:
#create empty dataframes that will store your derived datasets

derived_dataset1 = pd.DataFrame(columns = ['Class', 'pos'])
derived_dataset2 = pd.DataFrame(columns=['Class', 'pos-np', 'entities'])

In [249]:
def get_pos(sentence, wanted_pos): #wanted_pos refers to the desired pos tagging
    verbs = []
    for token in sentence:
        if token.pos_ in wanted_pos:
            verbs.append(token.lemma_) # lemma returns a number. lemma_ return a string
    return ' '.join(verbs) # return value is as a string and not a list for countVectorizer

In [250]:
#As an example, we use the above function to fetch all the verbs. We store this information in our first derived dataset
derived_dataset1['pos'] = data['tokenized'].apply(lambda sent : get_pos(sent, ['VERB']))

# Make derived_dataset1 to have NOT_VERB and VERB
def get_pos(sentence, target_pos):
    doc = nlp(sentence)
    pos_tags = [token.pos_ for token in doc]
    return any(pos in target_pos for pos in pos_tags)
derived_dataset1['Class'] = data['Title'].apply(lambda sent: 'VERB' if get_pos(sent, ['VERB']) else 'NOT_VERB')

In [251]:
derived_dataset1.head()

Class                 pos
0  VERB  evacuate kill find
1  VERB        ask identify
2  VERB               issue
3  VERB      store find say
4  VERB  indict give wrench

In [252]:
#Change this line to fetch your desired pos taggings for the second derived dataset
#For Derived Dataset 2, you also need to include Named Entities
#Below is just an example of obtaining such entities on a specific sentence, but you would do NER
#on the dataset of your choice.
#You can choose the types of entities (dates, organization, people) that you want,
#and then in your derived dataset, just make sure you include these entities separated by spaces (as shown for verbs)
#in a previous cell.

# sentence = "apple is looking at buying U.K. startup for $1 billion"
# doc = nlp(sentence)

# for ent in doc.ents:
#     print(ent.text, ent.start_char, ent.end_char, ent.label_)

def get_entities(sentence): #Rework from the code given above by getting entities with their respecting field.
    doc = nlp(sentence)
    entities = []
    for ent in doc.ents:
        if ent.label_ in ['ORG', 'MONEY', 'DATE']: 
            entities.append((ent.text, ent.start_char, ent.end_char, ent.label_))
    return entities
def get_class(entity):
    return entity[3] if entity else None #Returning the Entitiy Type and put it into Class.

# Initializing derived_dataset 2.
derived_dataset2['pos-np'] = data['tokenized'].apply(lambda sent: get_pos(sent, ['NOUN']))
derived_dataset2['entities'] = data['tokenized'].apply(get_entities)
derived_dataset2['Class'] = derived_dataset2['entities'].apply(lambda entities: get_class(entities[0]) if entities else None)
derived_dataset2 = derived_dataset2[derived_dataset2['entities'].apply(len) > 0]
derived_dataset2 = derived_dataset2.reset_index(drop=True) #Drop blank entitiy types.
derived_dataset2.head()

Class  pos-np                                           entities
0   ORG    True                              [(OSHA, 93, 97, ORG)]
1   ORG    True           [(Morgan State University, 70, 93, ORG)]
2  DATE    True                       [(1-year-old, 67, 77, DATE)]
3   ORG    True                            [(Shakur, 17, 23, ORG)]
4  DATE    True  [(Most days, 0, 9, DATE), (June 3, 76, 82, DATE)]

Now that you have your derived datasets, you can move to perform your classificaton task.

**Perform Classification Task**

1. Encode the text as input features with associated values

- Encoding text into input features where it removes stopwords and getting tf values.

In [253]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.feature_extraction import text
stop_words = text.ENGLISH_STOP_WORDS 
data['Title'] = data['Title'].apply(lambda x: ' '.join([word for word in x.split() if word.lower() not in stop_words])) #Filter out stop words
count_vectorizer = CountVectorizer()
X_count = count_vectorizer.fit_transform(data['Title'])
tf_vectorizer = TfidfVectorizer(use_idf=False)
X_tf = tf_vectorizer.fit_transform(data['Title'])
print(X_tf) #tf values for references.

  (0, 711)	0.30151134457776363
  (0, 1588)	0.30151134457776363
  (0, 1243)	0.30151134457776363
  (0, 1515)	0.30151134457776363
  (0, 957)	0.30151134457776363
  (0, 4834)	0.30151134457776363
  (0, 3186)	0.30151134457776363
  (0, 1561)	0.30151134457776363
  (0, 2404)	0.30151134457776363
  (0, 3090)	0.30151134457776363
  (0, 1690)	0.30151134457776363
  (1, 427)	0.2773500981126146
  (1, 3282)	0.2773500981126146
  (1, 344)	0.2773500981126146
  (1, 3418)	0.2773500981126146
  (1, 2037)	0.2773500981126146
  (1, 2152)	0.2773500981126146
  (1, 3936)	0.2773500981126146
  (1, 3108)	0.2773500981126146
  (1, 2852)	0.2773500981126146
  (1, 4129)	0.2773500981126146
  (1, 4592)	0.2773500981126146
  (1, 2082)	0.2773500981126146
  (1, 1518)	0.2773500981126146
  (2, 327)	0.3333333333333333
  :	:
  (1408, 1957)	0.3779644730092272
  (1408, 139)	0.3779644730092272
  (1409, 3791)	0.3779644730092272
  (1409, 2948)	0.3779644730092272
  (1409, 2044)	0.3779644730092272
  (1409, 4199)	0.3779644730092272
  (1409, 2

2. Define 2 models using some default parameters

- Logisitc Regression Model

In [258]:
from sklearn.metrics import classification_report
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

#Derived Data
X2 = derived_dataset2['pos-np']
y2 = derived_dataset2['Class']

vectorizer = CountVectorizer()
X2 = vectorizer.fit_transform(X2)

X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.2, random_state=42) #Train Test Split Function

model2 = LogisticRegression() #Creating the Logistic Regression Model
model2.fit(X2_train, y2_train)
y2_pred = model2.predict(X2_test)
print("Logistic Regression Classification Report for derived_dataset2:")
print(classification_report(y2_test, y2_pred))

y = derived_dataset1['Class']
X_train, X_test, y_train, y_test = train_test_split(X_tf, y, test_size=0.2, random_state=42)

# Create a logistic regression model
model = LogisticRegression()

# Train the model
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Print the classification report
print("Logistic Regression Classification Report :")
print(classification_report(y_test, y_pred))


AttributeError: 'bool' object has no attribute 'lower'

In [ ]:
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import classification_report

#Derived Data
X = derived_dataset2['pos-np']
y = derived_dataset2['Class']

vectorizer = CountVectorizer()
X = vectorizer.fit_transform(X)
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) #Train Test Split Function

#Creating MLP Model
mlp_classifier = MLPClassifier() #MLPClassifier Model
mlp_classifier.fit(X_train, y_train)
y_pred = mlp_classifier.predict(X_test)

print("MLP Model Classification Report for Derived_dataset1:")
print(classification_report(y_test, y_pred))

#Derived Data
X = derived_dataset1['pos']
y = derived_dataset1['Class']

vectorizer = CountVectorizer()
X = vectorizer.fit_transform(X)
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(y)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) #Train Test Split Function

#Creating MLP Model
mlp_classifier = MLPClassifier() #MLPClassifier Model
mlp_classifier.fit(X_train, y_train)
y_pred = mlp_classifier.predict(X_test)

print("MLP Model Classification Report for Derived_dataset1:")
print(classification_report(y_test, y_pred))

MLP Model Classification Report for Derived_dataset1:
              precision    recall  f1-score   support

           0       0.70      0.54      0.61        35
           1       0.00      0.00      0.00         1
           2       0.83      0.91      0.87        91

    accuracy                           0.80       127
   macro avg       0.51      0.48      0.49       127
weighted avg       0.79      0.80      0.79       127

MLP Model Classification Report for Derived_dataset1:
              precision    recall  f1-score   support

           0       0.40      0.65      0.49        26
           1       0.96      0.90      0.93       257

    accuracy                           0.88       283
   macro avg       0.68      0.78      0.71       283
weighted avg       0.91      0.88      0.89       283



3. Train/Test/Evaluate models on 3 datasets

In [ ]:
from sklearn.metrics import precision_score, recall_score
from sklearn.model_selection import StratifiedKFold

y = derived_dataset2['Class'] #Target variable
stratified_kfold = StratifiedKFold(n_splits=4, shuffle=True, random_state=42)

# Logistic Regression Model Evaluation
logistic_precision_micro = []
logistic_recall_micro = []
logistic_precision_macro = []
logistic_recall_macro = []

for train_index, test_index in stratified_kfold.split(X, y):
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]

    model2.fit(X_train, y_train)
    y_pred = model2.predict(X_test)

    # Calculate precision and recall
    precision_micro = precision_score(y_test, y_pred, average='micro')
    recall_micro = recall_score(y_test, y_pred, average='micro')
    precision_macro = precision_score(y_test, y_pred, average='macro')
    recall_macro = recall_score(y_test, y_pred, average='macro')

    logistic_precision_micro.append(precision_micro)
    logistic_recall_micro.append(recall_micro)
    logistic_precision_macro.append(precision_macro)
    logistic_recall_macro.append(recall_macro)

# Calculate average precision and recall
avg_logistic_precision_micro = sum(logistic_precision_micro) / len(logistic_precision_micro)
avg_logistic_recall_micro = sum(logistic_recall_micro) / len(logistic_recall_micro)
avg_logistic_precision_macro = sum(logistic_precision_macro) / len(logistic_precision_macro)
avg_logistic_recall_macro = sum(logistic_recall_macro) / len(logistic_recall_macro)

# Print the results
print("Logistic Regression Model for Derived_Dataset1:")
print("Micro Precision:", avg_logistic_precision_micro)
print("Micro Recall:", avg_logistic_recall_micro)
print("Macro Precision:", avg_logistic_precision_macro)
print("Macro Recall:", avg_logistic_recall_macro)



Logistic Regression Model for Derived_Dataset1:
Micro Precision: 0.8987252124645893
Micro Recall: 0.8987252124645893
Macro Precision: 0.4493626062322946
Macro Recall: 0.5


In [ ]:
from sklearn.metrics import precision_score, recall_score
from sklearn.model_selection import StratifiedKFold

y = derived_dataset2['Class'] #Target variable
stratified_kfold = StratifiedKFold(n_splits=4, shuffle=True, random_state=42)

# Logistic Regression Model Evaluation
logistic_precision_micro = []
logistic_recall_micro = []
logistic_precision_macro = []
logistic_recall_macro = []

for train_index, test_index in stratified_kfold.split(X, y):
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]

    model2.fit(X_train, y_train)
    y_pred = model2.predict(X_test)

    # Calculate precision and recall
    precision_micro = precision_score(y_test, y_pred, average='micro')
    recall_micro = recall_score(y_test, y_pred, average='micro')
    precision_macro = precision_score(y_test, y_pred, average='macro')
    recall_macro = recall_score(y_test, y_pred, average='macro')

    logistic_precision_micro.append(precision_micro)
    logistic_recall_micro.append(recall_micro)
    logistic_precision_macro.append(precision_macro)
    logistic_recall_macro.append(recall_macro)

# Calculate average precision and recall
avg_logistic_precision_micro = sum(logistic_precision_micro) / len(logistic_precision_micro)
avg_logistic_recall_micro = sum(logistic_recall_micro) / len(logistic_recall_micro)
avg_logistic_precision_macro = sum(logistic_precision_macro) / len(logistic_precision_macro)
avg_logistic_recall_macro = sum(logistic_recall_macro) / len(logistic_recall_macro)

# Print the results
print("Logistic Regression Model:")
print("Micro Precision:", avg_logistic_precision_micro)
print("Micro Recall:", avg_logistic_recall_micro)
print("Macro Precision:", avg_logistic_precision_macro)
print("Macro Recall:", avg_logistic_recall_macro)



ValueError: Found input variables with inconsistent numbers of samples: [1412, 634]

In [ ]:

# MLP Classifier Model Evaluation
mlp_precision_micro = []
mlp_recall_micro = []
mlp_precision_macro = []
mlp_recall_macro = []

for train_index, test_index in stratified_kfold.split(X, y):
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]

    mlp_classifier.fit(X_train, y_train)
    y_pred = mlp_classifier.predict(X_test)

    # Calculate precision and recall
    precision_micro = precision_score(y_test, y_pred, average='micro')
    recall_micro = recall_score(y_test, y_pred, average='micro')
    precision_macro = precision_score(y_test, y_pred, average='macro')
    recall_macro = recall_score(y_test, y_pred, average='macro')

    mlp_precision_micro.append(precision_micro)
    mlp_recall_micro.append(recall_micro)
    mlp_precision_macro.append(precision_macro)
    mlp_recall_macro.append(recall_macro)
    
# Calculate average precision and recall
avg_mlp_precision_micro = sum(mlp_precision_micro) / len(mlp_precision_micro)
avg_mlp_recall_micro = sum(mlp_recall_micro) / len(mlp_recall_micro)
avg_mlp_precision_macro = sum(mlp_precision_macro) / len(mlp_precision_macro)
avg_mlp_recall_macro = sum(mlp_recall_macro) / len(mlp_recall_macro)
    
print("MLP Classifier Model:")
print("Micro Precision:", avg_mlp_precision_micro)
print("Micro Recall:", avg_mlp_recall_micro)
print("Macro Precision:", avg_mlp_precision_macro)
print("Macro Recall:", avg_mlp_recall_macro)

MLP Classifier Model:
Micro Precision: 0.7854967757344161
Micro Recall: 0.7854967757344161
Macro Precision: 0.48835236715800223
Macro Recall: 0.47708267443086333


4. Modify parameters of MLP model and perform train/test/evaluate.

In [ ]:
y = derived_dataset2['Class']

# StratifiedKFold to ensure class balance in each fold
stratified_kfold = StratifiedKFold(n_splits=4, shuffle=True, random_state=42) #4 Fold Cross Validation

# Test Case 1
mlp_new1_precision_micro = []
mlp_new1_recall_micro = []
mlp_new1_precision_macro = []
mlp_new1_recall_macro = []

# Modify hidden_layer_sizes and activation parameters
# hidden_layer_sizesarray-like of shape(n_layers - 2,), default=(100,)
# ‘logistic’, the logistic sigmoid function, returns f(x) = 1 / (1 + exp(-x)).
# max_iterint, default=200 | Maximum number of iterations. The solver iterates until convergence (determined by ‘tol’) or this number of iterations.
# random_stateint, RandomState instance, default=None | Determines random number generation for weights and bias initialization, train-test split if early stopping is used, and batch sampling when solver=’sgd’ or ‘adam’.
mlp_model_new1 = MLPClassifier(hidden_layer_sizes=(25), activation='logistic', max_iter=500, random_state=42)

for train_index, test_index in stratified_kfold.split(X, y):
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]

    mlp_model_new1.fit(X_train, y_train)
    y_pred = mlp_model_new1.predict(X_test)

    # Calculate precision and recall
    precision_micro = precision_score(y_test, y_pred, average='micro')
    recall_micro = recall_score(y_test, y_pred, average='micro')
    precision_macro = precision_score(y_test, y_pred, average='macro')
    recall_macro = recall_score(y_test, y_pred, average='macro')

    mlp_new1_precision_micro.append(precision_micro)
    mlp_new1_recall_micro.append(recall_micro)
    mlp_new1_precision_macro.append(precision_macro)
    mlp_new1_recall_macro.append(recall_macro)

# Calculate average with new parameters
avg_mlp_new1_precision_micro = sum(mlp_new1_precision_micro) / len(mlp_new1_precision_micro)
avg_mlp_new1_recall_micro = sum(mlp_new1_recall_micro) / len(mlp_new1_recall_micro)
avg_mlp_new1_precision_macro = sum(mlp_new1_precision_macro) / len(mlp_new1_precision_macro)
avg_mlp_new1_recall_macro = sum(mlp_new1_recall_macro) / len(mlp_new1_recall_macro)

print("MLP Classifier Model w/ New Params:")
print("Micro Precision:", avg_mlp_new1_precision_micro)
print("Micro Recall:", avg_mlp_new1_recall_micro)
print("Macro Precision:", avg_mlp_new1_precision_macro)
print("Macro Recall:", avg_mlp_new1_recall_macro)


MLP Classifier Model w/ New Params:
Micro Precision: 0.7649868641031765
Micro Recall: 0.7649868641031765
Macro Precision: 0.4668494434869018
Macro Recall: 0.46512917589130126


In [ ]:
# Test Case 2
mlp_new2_precision_micro = []
mlp_new2_recall_micro = []
mlp_new2_precision_macro = []
mlp_new2_recall_macro = []

stratified_kfold = StratifiedKFold(n_splits=4, shuffle=True, random_state=69) #4 Fold Cross Validation

# learning_rate_initfloat, default=0.001 | The initial learning rate used. It controls the step-size in updating the weights. Only used when solver=’sgd’ or ‘adam’.
# max_iterint, default=200 | Maximum number of iterations. The solver iterates until convergence (determined by ‘tol’) or this number of iterations.
# random_stateint, RandomState instance, default=None | Determines random number generation for weights and bias initialization, train-test split if early stopping is used, and batch sampling when solver=’sgd’ or ‘adam’.
mlp_model_new2 = MLPClassifier(learning_rate_init=0.001, max_iter=420, random_state=69)

for train_index, test_index in stratified_kfold.split(X, y):
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]

    mlp_model_new2.fit(X_train, y_train)
    y_pred = mlp_model_new2.predict(X_test) 

    # Calculate precision and recall
    precision_micro = precision_score(y_test, y_pred, average='micro')
    recall_micro = recall_score(y_test, y_pred, average='micro')
    precision_macro = precision_score(y_test, y_pred, average='macro')
    recall_macro = recall_score(y_test, y_pred, average='macro')

    mlp_new2_precision_micro.append(precision_micro)
    mlp_new2_recall_micro.append(recall_micro)
    mlp_new2_precision_macro.append(precision_macro)
    mlp_new2_recall_macro.append(recall_macro)

# Calculate with new parameters
avg_mlp_new2_precision_micro = sum(mlp_new2_precision_micro) / len(mlp_new2_precision_micro)
avg_mlp_new2_recall_micro = sum(mlp_new2_recall_micro) / len(mlp_new2_recall_micro)
avg_mlp_new2_precision_macro = sum(mlp_new2_precision_macro) / len(mlp_new2_precision_macro)
avg_mlp_new2_recall_macro = sum(mlp_new2_recall_macro) / len(mlp_new2_recall_macro)

print("MLP Classifier Model w/ Params 2:")
print("Micro Precision:", avg_mlp_new2_precision_micro)
print("Micro Recall:", avg_mlp_new2_recall_micro)
print("Macro Precision:", avg_mlp_new2_precision_macro)
print("Macro Recall:", avg_mlp_new2_recall_macro)


MLP Classifier Model w/ Params 2:
Micro Precision: 0.7855266300453785
Micro Recall: 0.7855266300453785
Macro Precision: 0.49119806119417775
Macro Recall: 0.47703186142273324


5. Documentation

Logistic Regression Classification Report:
              precision    recall  f1-score   support

        DATE       0.73      0.46      0.56        35
       MONEY       0.00      0.00      0.00         1
         ORG       0.81      0.93      0.87        91

    accuracy                           0.80       127
   macro avg       0.51      0.46      0.48       127
weighted avg       0.78      0.80      0.78       127

MLP Model Classification Report:
              precision    recall  f1-score   support

           0       0.69      0.51      0.59        35
           1       0.00      0.00      0.00         1
           2       0.82      0.91      0.86        91

    accuracy                           0.80       127
   macro avg       0.50      0.48      0.48       127
weighted avg       0.78      0.80      0.78       127

Logistic Regression Model:
Micro Precision: 0.8075690629726933
Micro Recall: 0.8075690629726933
Macro Precision: 0.5256874461751904
Macro Recall: 0.4807136533893943

MLP Classifier Model:
Micro Precision: 0.7854967757344161
Micro Recall: 0.7854967757344161
Macro Precision: 0.48835236715800223
Macro Recall: 0.47708267443086333

MLP Classifier Model w/ New Params:
Micro Precision: 0.7649868641031765
Micro Recall: 0.7649868641031765
Macro Precision: 0.4668494434869018
Macro Recall: 0.46512917589130126

MLP Classifier Model w/ Params 2:
Micro Precision: 0.7855266300453785
Micro Recall: 0.7855266300453785
Macro Precision: 0.49119806119417775
Macro Recall: 0.47703186142273324

Reference:
- https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html
- https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html
- https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression
- https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedKFold.html
- https://scikit-learn.org/stable/modules/cross_validation.html
- https://www.geeksforgeeks.org/python-extracting-rows-using-pandas-iloc/
- https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html
- https://stackoverflow.com/questions/72348496/my-k-fold-cross-validation-technique-is-giving-error-on-my-dataframe-with-delete
